crates/bevy_ecs/src/system/async_system.rs

```diff
@@ -0,0 +1,240 @@
+use std::{any::TypeId, borrow::Cow, ops::Deref};
+
+use async_channel::{Receiver, Sender};
+
+use bevy_tasks::TaskPool;
+use bevy_utils::BoxedFuture;
+
+use crate::{
+    ArchetypeComponent, FetchSystemParam, Resources, System, SystemId, SystemParam,
+    SystemState, TypeAccess, World,
+};
+
+pub trait AsyncSystem<Params: SystemParam>: Send + Sync + Copy + 'static {
+    type TaskPool: Deref<Target = TaskPool> + Send + Sync;
+
+    fn run(self, accessor: Accessor<Params>) -> BoxedFuture<'static, ()>;
+
+    fn system(self) -> AsyncSystemContainer<Params, Self>;
+}
+
+impl<P, F, Func> AsyncSystem<P> for Func
+where
+    P: SystemParam + 'static,
+    F: std::future::Future<Output = ()> + Send + Sync + 'static,
+    Func: Fn(Accessor<P>) -> F + Send + Sync + Copy + 'static,
+{
+    type TaskPool = bevy_tasks::AsyncComputeTaskPool;
+
+    fn run(self, accessor: Accessor<P>) -> BoxedFuture<'static, ()> {
+        Box::pin((self)(accessor))
+    }
+
+    fn system(self) -> AsyncSystemContainer<P, Self> {
+        let mut resource_access = TypeAccess::default();
+        resource_access.add_write(TypeId::of::<Self::TaskPool>());
+
+        AsyncSystemContainer::<P, Self> {
+            state: SystemState {
+                name: std::any::type_name::<Self>().into(),
+                archetype_component_access: TypeAccess::default(),
+                resource_access,
+                local_resource_access: TypeAccess::default(),
+                id: SystemId::new(),
+                commands: Default::default(),
+                arc_commands: Default::default(),
+                current_query_index: Default::default(),
+                query_archetype_component_accesses: Vec::new(),
+                query_accesses: Vec::new(),
+                query_type_names: Vec::new(),
+            },
+            channel: async_channel::bounded(1).1,
+            core: self,
+        }
+    }
+}
+
+pub struct Accessor<P: SystemParam> {
+    channel: Sender<Box<dyn FnOnce(<P::Fetch as FetchSystemParam<'static>>::Item) + Send + Sync>>,
+}
+
+impl<P: SystemParam> Accessor<P> {
+    pub async fn access<F, R: Send + 'static>(&mut self, sync: F) -> R
+    where
+        F: FnOnce(<P::Fetch as FetchSystemParam<'static>>::Item) -> R + Send + Sync + 'static,
```

Comment:
So this line is the problem. One approach would be to send `P::Fetch` in the channel, and then do the actual fetching here.

Reply:
That is, force `F` to be `for <'a> FnOnce(<P::Fetch as FetchSystemParam<'a>>::Item)`